cudax/include/cuda/experimental/__memory_resource/async_memory_pool.h

```diff
@@ -0,0 +1,306 @@
+  explicit async_memory_pool(const ::cuda::experimental::device_ref __device_id,
+                             async_memory_pool_properties __properties = {})
+      : __pool_handle_(__create_cuda_mempool(__device_id.get(), __properties))
+  {}
+
+  //! @brief Disables construction from a plain `cudaMemPool_t`. We want to ensure clean ownership semantics.
+  async_memory_pool(::cudaMemPool_t) = delete;
+
+  async_memory_pool(async_memory_pool const&)            = delete;
+  async_memory_pool(async_memory_pool&&)                 = delete;
+  async_memory_pool& operator=(async_memory_pool const&) = delete;
+  async_memory_pool& operator=(async_memory_pool&&)      = delete;
+
+  //! @brief Destroys the \c async_memory_pool by releasing the internal ``cudaMemPool_t``.
+  ~async_memory_pool() noexcept
+  {
+    _CCCL_ASSERT_CUDA_API(::cudaMemPoolDestroy, "~async_memory_pool() failed to destroy pool", __pool_handle_);
+  }
+
+  //! @brief Equality comparison with another \c async_memory_pool.
+  //! @returns true if the stored ``cudaMemPool_t`` are equal.
+  _CCCL_NODISCARD constexpr bool operator==(async_memory_pool const& __rhs) const noexcept
+  {
+    return __pool_handle_ == __rhs.__pool_handle_;
+  }
+
+#    if _CCCL_STD_VER <= 2017
+  //! @brief Inequality comparison with another \c async_memory_pool.
+  //! @returns true if the stored ``cudaMemPool_t`` are not equal.
+  _CCCL_NODISCARD constexpr bool operator!=(async_memory_pool const& __rhs) const noexcept
+  {
+    return __pool_handle_ != __rhs.__pool_handle_;
+  }
+#    endif // _CCCL_STD_VER <= 2017
+
+  //! @brief Equality comparison with a \c cudaMemPool_t.
+  //! @param __rhs A \c cudaMemPool_t.
+  //! @returns true if the stored ``cudaMemPool_t`` is equal to \p __rhs.
+  _CCCL_NODISCARD_FRIEND constexpr bool operator==(async_memory_pool const& __lhs, ::cudaMemPool_t __rhs) noexcept
+  {
+    return __lhs.__pool_handle_ == __rhs;
+  }
+
+#    if _CCCL_STD_VER <= 2017
+  //! @copydoc async_memory_pool::operator==(async_memory_pool const&, ::cudaMemPool_t)
+  _CCCL_NODISCARD_FRIEND constexpr bool operator==(::cudaMemPool_t __lhs, async_memory_pool const& __rhs) noexcept
+  {
+    return __rhs.__pool_handle_ == __lhs;
+  }
+
+  //! @copydoc async_memory_pool::operator==(async_memory_pool const&, ::cudaMemPool_t)
+  _CCCL_NODISCARD_FRIEND constexpr bool operator!=(async_memory_pool const& __lhs, ::cudaMemPool_t __rhs) noexcept
+  {
+    return __lhs.__pool_handle_ != __rhs;
+  }
+
+  //! @copydoc async_memory_pool::operator==(async_memory_pool const&, ::cudaMemPool_t)
+  _CCCL_NODISCARD_FRIEND constexpr bool operator!=(::cudaMemPool_t __lhs, async_memory_pool const& __rhs) noexcept
+  {
+    return __rhs.__pool_handle_ != __lhs;
+  }
+#    endif // _CCCL_STD_VER <= 2017
+
+  //! @brief Returns the underlying handle to the CUDA memory pool.
+  _CCCL_NODISCARD constexpr cudaMemPool_t get() const noexcept
+  {
+    return __pool_handle_;
+  }
+
```

Comment:
There should be a `release()` member to relinquish ownership of the handle.

Reply:
I am not convinced that this is something we currently want. We can always add it later